### ISSUE TITLE
Segfault computing `SylowSubgroup(G,12);`

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#1844]
The following snippet causes `master` to segfault:
```gap
G:=SymmetricGroup(5); SylowSubgroup(G,12);
```
Using `git bisect`, I tracked this down to 771e34b44a044a7b97427cb6af8b38ca8ed63caa

However, I just stared at this commit for half an hour and can't figure out how that might cause this breakage. But it definitely does do so...

**Comment by markuspf** [comment]
Max Horn writes:

> The following snippet causes `master` to segfault:
> ```gap
> G:=SymmetricGroup(5); SylowSubgroup(G,12);
> ```
> Using `git bisect`, I tracked this down to 771e34b44a044a7b97427cb6af8b38ca8ed63caa
>
> However, I just stared at this commit for half an hour and can't figure out how that might cause this breakage. But it definitely does do so...

I think this commit only exposes the problem: If you checkout the one
immediately previously the error output is incomplete already (no file
number, and the line number is suspicious, too). I suspect it has
something to do with this being a function expression inside a function
(the KeyDependent installation stuff).

I am too tired right now to stare more though...

**Comment by fingolfin** [comment]
OK, I think I am getting somewhere:

Backtrace:
```
    frame #0: 0x00000001000ab615 gap`TNUM_STAT(stat=17592186750336) + 21 at code.h:244
    frame #1: 0x00000001000b0909 gap`PrintExpr(expr=17592186750336) + 89 at exprs.c:1453
    frame #2: 0x00000001000c150b gap`PrintFunccall1(call=5672) + 59 at funcs.c:1661
    frame #3: 0x00000001000c13e5 gap`PrintFunccall(call=5672) + 21 at funcs.c:1678
    frame #4: 0x00000001000c13a5 gap`PrintProccall(call=5672) + 21 at funcs.c:1636
    frame #5: 0x00000001001b8806 gap`PrintStat(stat=5672) + 38 at stats.c:1822
  * frame #6: 0x00000001000c608d gap`FuncPRINT_CURRENT_STATEMENT(self=0x00001000000101e0, context=0x0000100000500230) + 173 at gap.c:1072
    frame #7: 0x00000001000b9473 gap`ExecProccall1args(call=5232) + 867 at funcs.c:210
    frame #8: 0x00000001001b2cbb gap`EXEC_STAT(stat=5232) + 59 at stats.h:54
    frame #9: 0x00000001001b2e4a gap`ExecSeqStat4(stat=5432) + 90 at stats.c:224
    frame #10: 0x00000001001b2cbb gap`EXEC_STAT(stat=5432) + 59 at stats.h:54
    frame #11: 0x00000001001b347c gap`ExecIfElse(stat=5584) + 220 at stats.c:333
    frame #12: 0x00000001001b2cbb gap`EXEC_STAT(stat=5584) + 59 at stats.h:54
    frame #13: 0x00000001001b347c gap`ExecIfElse(stat=6104) + 220 at stats.c:333
    frame #14: 0x00000001001b2cbb gap`EXEC_STAT(stat=6104) + 59 at stats.h:54
    frame #15: 0x00000001001b2c0c gap`ExecSeqStat(stat=8104) + 92 at stats.c:182
    frame #16: 0x00000001001b2cbb gap`EXEC_STAT(stat=8104) + 59 at stats.h:54
    frame #17: 0x00000001001b3283 gap`ExecSeqStat7(stat=32) + 355 at stats.c:276
    frame #18: 0x00000001000b8ccb gap`EXEC_STAT(stat=32) + 59 at stats.h:54
    frame #19: 0x00000001000c0184 gap`ExecFuncHelper + 36 at funcs.c:855
    frame #20: 0x00000001000c035c gap`DoExecFunc1args(func=0x0000100000077db8, arg1=0x0000100000500308) + 108 at funcs.c:906
    frame #21: 0x0000000100072b4c gap`DoWrap2args(self=0x0000100000077db8, arg1=0x00001000005002f0, arg2=0x0000100000500300) + 140 at calls.c:161
    frame #22: 0x00000001000c6a82 gap`CallErrorInner(msg="List Element: <list>[%d] must have an assigned value", arg1=22600, arg2=0, justQuit=0, mayReturnVoid=1, mayReturnObj=0, lateMessage=0x00001000005002e8, printThisStatement=1) + 850 at gap.c:1259
    frame #23: 0x00000001000c6e9b gap`ErrorReturnVoid(msg="List Element: <list>[%d] must have an assigned value", arg1=22600, arg2=0, msg2="you can 'return;' after assigning a value") + 91 at gap.c:1435
    frame #24: 0x00000001001679ce gap`ElmPlist(list=0x0000100000002420, pos=22600) + 62 at plist.c:1260
    frame #25: 0x0000000100077f1e gap`ELM_LIST(list=0x0000100000002420, pos=22600) + 46 at lists.h:325
    frame #26: 0x0000000100077ea6 gap`FILENAME_STAT(stat=0) + 70 at code.c:145
    frame #27: 0x00000001000c5e3e gap`FuncCURRENT_STATEMENT_LOCATION(self=0x0000100000010220, context=0x0000100000500230) + 382 at gap.c:1048
    frame #28: 0x00000001000bcef8 gap`EvalFunccall1args(call=5648) + 840 at funcs.c:516
    frame #29: 0x00000001001eb18c gap`ExecAssLVar(stat=5672) + 364 at vars.c:155
    frame #30: 0x00000001001b2cbb gap`EXEC_STAT(stat=5672) + 59 at stats.h:54
    frame #31: 0x00000001001b2d75 gap`ExecSeqStat3(stat=6072) + 37 at stats.c:210
    frame #32: 0x00000001001b2cbb gap`EXEC_STAT(stat=6072) + 59 at stats.h:54```
Looking a bit up, we see this in frame #6:
```
frame #6: 0x00000001000c608d gap`FuncPRINT_CURRENT_STATEMENT(self=0x00001000000101e0, context=0x0000100000500230) + 173 at gap.c:1072
   1069	  }
   1070	    else if ( /*FIRST_STAT_TNUM <= TNUM_STAT(call) &&*/
   1071	              TNUM_STAT(call) <= LAST_STAT_TNUM ) {
-> 1072	      PrintStat( call );
   1073	      Pr(" at %s:%d",(UInt)CSTR_STRING(FILENAME_STAT(call)),LINE_STAT(call));
   1074	    }
   1075	    else if ( FIRST_EXPR_TNUM <= TNUM_EXPR(call) &&
```
GAP thinks we are looking a statement at offset 5672 (the value of `call`), but the current function bag is only 24 bytes large:
```
(lldb) p SIZE_BAG(BODY_FUNC(CURR_FUNC()))
(UInt) $13 = 24
(lldb) p call
(Expr) $14 = 5672
```

So the crash is caused by `FuncPRINT_CURRENT_STATEMENT` seeing / using an invalid stat offset (we should probably add `GAP_ASSERT` bounds checks for that in various places, to catch similar problems sooner in the future).

Anyway, the bad offset is 5672, and looking further back in the backtrace, we see were it comes from: it was the active statement prior to a call to `FuncCURRENT_STATEMENT_LOCATION` in frame #27! That call also saw some invalid data, namely an invalid fileid 22600 inside a `StatHeader`.

So, there are several issues that come together here:

1. The root cause (as discovered so far) is a statement with an invalid fileid (which likely mans that the stat offset at that point also was invalid?)
2. We generally don't perform bounds checks (or any other sanity checks) on statement offsets and/or the values inside `StatHeaders`; we should do that!
3. The system is not prepared for an error inside `FuncCURRENT_STATEMENT_LOCATION`: as that function modifies the active frame via `SWITCH_TO_OLD_LVARS`, really bad things can happen here -- e.g. `BRK_CALL_TO` gets messed up by `CallErrorInner`

I'll dig some more...

**Comment by fingolfin** [comment]
@markuspf aha, good observation, thanks

**Comment by fingolfin** [comment]
Ahhh, indeed: the function in which the error occurred is this one:
```
      keytest := function( key )
          if not IsPrimeInt( key ) then
            Error( name, ": <p> must be a prime" );
          fi;
      end;
```
But that's in `oper1.g`, so compiled, so it's really `HdlrFunc14` from `c_oper1.g` -- and that's precisely the active function bag at that point.

Now we should be showing the "compiled statement" message, but we actually don't have a proper check for that in place inside of `PRINT_CURRENT_STATEMENT` (and `CURRENT_STATEMENT_LOCATION`), hence the crash.

So, basically just have to add a check for "this is a compiled function" to fix the crash. Will look into that tomorrow.
